spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/functions/TruncateFunction.java

```
@@ -0,0 +1,319 @@
+          String.format("Invalid input type. Expected 2 fields but found %s", inputType.fields().length));
+    }
+
+    StructField fieldToTruncate = inputType.apply(0);
+    StructField truncationWidth = inputType.apply(1);
+
+    DataType toTruncateSparkType = fieldToTruncate.dataType();
+    Type sourceIcebergType = SparkSchemaUtil.convert(fieldToTruncate.dataType());
+
+    validateSourceTypeUsingSparkTypes(fieldToTruncate.dataType());
+    validateTruncationWidthType(truncationWidth.dataType());
+
+    switch (sourceIcebergType.typeId()) {
+      case INTEGER:
+        return new TruncateInteger();
+      case LONG:
+        return new TruncateLong();
+      case DECIMAL:
+        return new TruncateDecimal(
+            ((DecimalType) toTruncateSparkType).precision(),
+            ((DecimalType) toTruncateSparkType).scale());
+      case STRING:
+        return new TruncateString();
+      case BINARY:
+        return new TruncateBinary();
+      default:
+        throw new UnsupportedOperationException(
+            "Cannot truncate type: " + sourceIcebergType.typeId());
+    }
+  }
+
+  @Override
+  public String description() {
+    return "Truncate - The Iceberg truncate function used for truncate partition transformations.";
+  }
+
+  @Override
+  public String name() {
+    return "truncate";
+  }
+
+  public abstract static class TruncateBase<T> implements ScalarFunction<T> {
+    @Override
+    public String name() {
+      return "truncate";
+    }
+
+    @Override
+    public boolean isResultNullable() {
+      return true;
+    }
+
+    @Override
+    public boolean isDeterministic() {
+      return true;
+    }
+  }
+
+  public static class TruncateInteger extends TruncateBase<Integer> {
+    // magic method used in codegen
+    public static int invoke(int value, int width) {
+      if (width <= 0) {
+        throw new IllegalArgumentException(String.format("Invalid truncate width: %s (must be > 0)", width));
```

I probably wouldn't bother with this check.

---

Should we just let the code fail? For some calls, it seems the code would do things that don't necessarily work (based on the existing transforms).

---

It won't fail because these are primitives.

---

Buyt the width could be negative. The range for Integer is signed integer range. That's my concern (not so much that we'll get null for the width).

---

Yeah, if you use weird values, you get weird results.

---

I'll add a note in the JavaDoc and then remove that (as throwing exceptions in generated code is somewhat of an anti-pattern if possible).